### PR TITLE
Namespace is always zkat even if it was changed in the node config #256

### PR DESCRIPTION
**Pull request by adecaro**
When invoking token.GetManagementService() without any additional option, the token-api tries to load the default TMS. What does this mean? The token-api checks if there is a default fabric or orion network configuration, if yes it consider that the default. For the channel, it is the same. For the namespace, the issue is that there can be multiple token namespaces in the same <network, channel>. So, the code currently uses the default namespace zkat.

So, the solution we can put forth is the following: The code checks the configuration and if there is a unique tuple <network, channel, namespace>, the code will use that one, otherwise it will return an error.

Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

File: integration/token/fungible/topology.go [hunk 1]
```diff
@@ -141,6 +141,7 @@ func Topology(backend string, tokenSDKDriver string) []api.Topology {
 	tokenTopology := token.NewTopology()
 	tokenTopology.SetDefaultSDK(fscTopology)
 	tms := tokenTopology.AddTMS(backendNetwork, backendChannel, tokenSDKDriver)
+	tms.Namespace = "token-chaincode"
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply integration/token/fungible/topology.go:144]: should we also add this to some FSCNode core.yaml template with some default value?

File: token/core/config/config.go [hunk 2]
```diff
@@ -41,10 +44,51 @@ func NewManager(cp configProvider, network, channel, namespace string) (*Manager
...
-func (m *Manager) TranslatePath(path string) string {
+// TranslatePath translates the passed path relative to the config path
+func (m *TMSProvider) TranslatePath(path string) string {
 	return m.cp.TranslatePath(path)
 }
+
+// Manager manages the configuration of the token-sdk.
+type Manager struct {
+	cp configProvider
+}
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply token/core/config/config.go:60]: So why do we need a manager and providers here? 

In other components in FSC and FTS we use providers to create a new objects/instances of a service (or sometimes return a singleton). A provider is usually "pre-configured" with some information, which is used to when the provider is called. A manager is a construct which helps as to deal with many instances or a particular component. How do these concepts apply for a TMS configuration? I don't see this at the moment. 

What do we need here?
- some functions to extract a subset of a configuration for a given network, channel, namspace tuple.
- if we don't know a namespace, perform a lookup in the configuration for a given network/channel tuple

Can't we implement this functionality with just a helper function without dealing with providers and managers ....

**Review by mbrandenburger — COMMENTED** [review]
Thanks Angelo

File: token/core/config/config.go [hunk 3]
```diff
@@ -41,10 +44,51 @@ func NewManager(cp configProvider, network, channel, namespace string) (*Manager
 	return nil, errors.Errorf("no token-sdk configuration for network %s, channel %s, namespace %s", network, channel, namespace)
 }
 
-func (m *Manager) TMS() *driver.TMS {
+// TMS returns the TMS configuration this provider is associated with.
+func (m *TMSProvider) TMS() *driver.TMS {
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/config/config.go:48]: I'll make more clear

File: token/core/config/config.go [hunk 4]
```diff
@@ -16,21 +16,24 @@ type configProvider interface {
 	TranslatePath(path string) string
 }
 
-type Manager struct {
+// TMSProvider is a config provider for a specific TMS
+type TMSProvider struct {
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply token/core/config/config.go:20]: true - I don't have a strong opinion here but I am just saying that the naming doesn't feel ideal to me.
- `core.TMSProvider` : tells me it returns a TMS
- `config.TMSProvider` :  tells me it returns a TMS in the context of some configuration?! (I am just guessing here and need to read more documentation)

File: integration/token/fungible/topology.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply integration/token/fungible/topology.go:144]: can one also specify the namespace via `core.yaml`? if yes, should we add this somewhere?

[the remark this answers is not in the document]
Comment (adecaro) [reply integration/token/fungible/topology.go]: that was already possible: https://github.com/hyperledger-labs/fabric-token-sdk/blob/main/samples/fungible/README.md#anatomy-of-the-configuration-file 
Should I add more about that? @mbrandenburger

[the remark this answers is not in the document]
Comment (mbrandenburger) [reply integration/token/fungible/topology.go]: aha! perfect then!

File: integration/nwo/token/topology/topology.go [hunk 5]
```diff
@@ -58,3 +80,23 @@ func (t *TMS) SetTokenGenPublicParams(publicParamsGenArgs ...string) {
...
+	// Check if namespace is valid
+	err := ValidateNs(namespace)
+	if err != nil {
+		panic(errors.WithMessagef(err, "invalid namespace '%s'", namespace))
+	}
+	// Check if another namespace already exists in the same <network,channel>
+	for _, tms := range t.TokenTopology.GetTMSs() {
+		if tms.Network == t.Network && tms.Channel == t.Channel {
+			if tms.Namespace == namespace {
+				panic(fmt.Sprintf("Namespace [%s] already exists", namespace))
```
[the remark this answers is not in the document]
Comment (adecaro) [reply integration/nwo/token/topology/topology.go:95]: 👍

File: token/core/config/config.go [hunk 6]
```diff
@@ -16,69 +16,43 @@ type configProvider interface {
...
 }
 
-// SearchTMS searches for a TMS configuration that matches the given network and channel, and
+// LookupNamespace searches for a TMS configuration that matches the given network and channel, and
 // return its namespace.
 // If no matching configuration is found, an error is returned.
 // If multiple matching configurations are found, an error is returned.
-func (m *Manager) SearchTMS(network, channel string) (string, error) {
-	var tmsConfigs []*driver.TMS
+func (m *TokenSDK) LookupNamespace(network, channel string) (string, error) {
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/config/config.go:49]: the problem is that I'm searching for a namespace, no? 🤔

**Pull request merged**